File: go/0088/solution.go
package solution

func merge(nums1 []int, m int, nums2 []int, n int) {
	curr := len(nums1) - 1
	m -= 1
	n -= 1
	for n >= 0 {
		if m >= 0 && nums1[m] > nums2[n] {
			nums1[curr] = nums1[m]
			m--
		} else {
			nums1[curr] = nums2[n]
			n--
		}
		curr--
	}
}

// O(n) time solve
/*
[1, 2, 3, 0, 0, 0], [2, 5, 6]

[1, 2, 3, 0, 0, 0]
       3                    6
[1, 2, 3, 0, 0, 6]
       3                5
[1, 2, 3, 0, 5, 6]
       3                2
[1, 2, 0, 3, 5, 6]
    2                   2
[1, 2, 2, 3, 5, 6]

move backwards, pointer m, n, curr
*/

File: cpp/0088/solution.cc
#include "solution.h"

void Solution::merge(std::vector<int> &nums1, int m, std::vector<int> &nums2,
                     int n) {
  int s = m + n - 1;
  m = m - 1;
  n = n - 1;
  while (s >= 0) {
    int first_vector = (m >= 0) ? nums1[m] : INT_MIN;
    int second_vector = (n >= 0) ? nums2[n] : INT_MIN;
    if (second_vector >= first_vector) {
      nums1[s] = second_vector;
      n -= 1;
    } else if (second_vector <= first_vector) {
      nums1[s] = first_vector;
      m -= 1;
    }
    s -= 1;
  }
}
